core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -593,4 +610,91 @@ public boolean isEmpty() {
+     * that incrementally reduces aggregation results as shard results are consumed.
+     * This implementation can be configured to batch up a certain amount of results and only reduce them
+     * iff the buffer is exhausted.
+     */
+    static class QueryPhaseResultConsumer
+        extends InitialSearchPhase.SearchPhaseResults<QuerySearchResultProvider> {
+        private final InternalAggregations[] buffer;
+        private int index;
+        private Boolean hasAggs = null;
+        private final SearchPhaseController controller;
+
+        /**
+         * Creates a new {@link QueryPhaseResultConsumer}
+         * @param controller a controller instance to reduce the query response objects
+         * @param expectedResultSize the expected number of query results. Corresponds to the number of shards queried
+         * @param bufferSize the size of the reduce buffer. if the buffer size is smaller than the number of expected results
+         *                   the buffer is used to incrementally reduce aggregation results before all shards responded.
+         */
+        QueryPhaseResultConsumer(SearchPhaseController controller, int expectedResultSize, int bufferSize) {
+            super(expectedResultSize);
+            this.controller = controller;
+            // TODO should we not allocate a buffer if there are no aggs in the search request?
+            if (expectedResultSize != 1 && bufferSize < 2) {
+                throw new IllegalArgumentException("buffer size must be >= 2 if there is more than one expected result");
+            }
+            // no need to buffer anything if we have less expected results. in this case we don't consume any results ahead of time.
+            this.buffer = expectedResultSize <= bufferSize ? null : new InternalAggregations[bufferSize];
+        }
+
+        @Override
+        public void consumeResult(int shardIndex, QuerySearchResultProvider result) {
+            super.consumeResult(shardIndex, result);
+            QuerySearchResult queryResult = result.queryResult();
+            if (queryResult.hasAggs()) {
+                consumeInternal(queryResult);
+            } else {
+                assert assertNoAggs();
+            }
+        }
+
+        private synchronized boolean assertNoAggs() {
+            assert hasAggs == null || hasAggs == false;
+            hasAggs = Boolean.FALSE;
+            return true;
+        }
+
+        private synchronized void consumeInternal(QuerySearchResult querySearchResult) {
+            assert hasAggs == null || hasAggs;
+            hasAggs = Boolean.TRUE;
+            if (buffer != null) {
+                InternalAggregations aggregations = (InternalAggregations) querySearchResult.consumeAggs();
+                // once the size is incremented to the length of the buffer we know all elements are added
+                // we also have happens before guarantees due to the memory barrier of the size write
```

Comment:
yeah I had a complex solution first with non-blocking concurrency etc. I didn't go with it apparently